### DISCUSSION TITLE
Getting all child nodes in treeView

### DISCUSSION THREAD

**Discussion by DmitryOffsec** [microsoft/vscode-discussions#886] [Extension Development QnA] [answered]
TreeView uses lazy retrieval of children, meaning that the `getChildren `method is not called until the parent node is open. Is it possible to somehow get all subnodes from the TreeView provider even if the user did not open the subnodes?
Thanks in advance!

**Comment by theReynald** [comment, marked as answer]
Hi @DmitryOffsec,

If you want to retrieve all nodes of a `TreeView`, including those not expanded by the user, you would have to programmatically expand all the nodes. However, this goes against the lazy-loading principle and could potentially lead to performance issues, especially if the tree is large.

VS Code does not provide a native API to forcibly retrieve all children of all nodes without user interaction. The `TreeView` API is designed around user interaction to drive the retrieval of data.

To achieve something similar to what you're asking, you would likely need to maintain your own data structure within your extension that represents the full tree. You could build and manage this structure independently of the `TreeView` and use it to access all nodes at any time.

Here's a high-level approach on how you might do this:

1. **Build a Complete Data Model:** When your extension is activated, construct a full data model of the tree. This model should reflect all possible nodes, regardless of whether they are currently visible in the `TreeView`.

2. **Synchronize with `TreeView`:** When `getChildren` is called by the `TreeView`, provide only the relevant subset of your complete data model that corresponds to the requested parent node.

3. **Access Subnodes:** Since you have a complete data model, you can access any node directly from your data model, not just those currently displayed in the `TreeView`.

It's important to note that while this method lets you access all nodes programmatically, it does not actually expand all nodes within the `TreeView` UI, which would still require user interaction. Additionally, you need to carefully manage the synchronization between your data model and the `TreeView` to ensure consistency and to handle updates.

If you need to perform operations on all nodes or need access to the entire structure for a specific purpose (like searching or processing data in the background), you would typically interact with your internal data model rather than using the `TreeView` API.

Hope this helps.
